storage: constrained span of rangedel in ClearRange to keys in range

OwenQian opened the pull request:
Constrains the width of the range deletion tombstone to the span of keys
actually present within the range. If the range has no kv-entries, then skip the
rangedel completely.

Before this change, when receiving a snapshot, the original file
would have a range deletion tombstone that spanned the entire range written to
it regardless of the actual keys contained in the range or if the range was
empty. This resulted in the creation of excessively wide tombstones, which has
significant performance implications since the wide tombstones impede
compaction.

Fixes #44048.

Rebased off #45100.

Release note: None.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44725)
<!-- Reviewable:end -->

nvb reviewed (commented):
Nice, thanks for getting this out @owenqian. We're going to want to add some testing that looks similar to `TestStoreRangeMergeRaftSnapshot` but tests a) the case where the recipient has an empty keyspace and b) the case where the recipient has a constrained set of keys in its keyspace. Finally, we'll want to add similar logic to `clearSubsumedReplicaDiskData`, where you left your TODO.

Also, if you add "Fixes #44048." to your commit message and/or PR message, Github will automatically link the PR to the issue and close the issue when we eventually merge the fix.

@sumeerbhola @petermattis @dt this should be far enough along to unblock you on #44048.

> Reviewed 1 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-:-M-HQluH0ehZExt-8pzt:b-v8u3h5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @OwenQian)

___
*[pkg/storage/replica_sst_snapshot_storage_test.go, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-HPBps991KlDhDKVjJ:-M-HPBpt11UCD_A0f4FO:b3wojet) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef6d8591152aabc7421b07a35ff3522e0dae53ec/pkg/storage/replica_sst_snapshot_storage_test.go#L52)):*
> ```Go
> 	}()
> 
> 	// Check that even though the files aren't created, they are still recorded in SSTs().
> ```

We'll need to update this comment. Then we'll want to test that once we start writing to the files, they get added to the slice.
___
*[pkg/storage/store_snapshot.go, line 110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-HPMPX05uySW-6M0aF:-M-HPMPX05uySW-6M0aG:b2eoe9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef6d8591152aabc7421b07a35ff3522e0dae53ec/pkg/storage/store_snapshot.go#L110)):*
> ```Go
> 	// Only used on the receiver side.
> 	scratch *SSTSnapshotStorageScratch
> 	// Used to restrict the span of the rangedel done in ClearRange when receiving a snapshot
> ```

Add a period to the end of this.
___
*[pkg/storage/store_snapshot.go, line 149 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-HPVg_A81bB8KLn33v:-M-HPVg_A81bB8KLn33w:bsvnx5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef6d8591152aabc7421b07a35ff3522e0dae53ec/pkg/storage/store_snapshot.go#L149)):*
> ```Go
> 
> func (msstw *multiSSTWriter) initSST(ctx context.Context) error {
> 	span := roachpb.Span{Key: msstw.keyRanges[msstw.currRange].Start.Key, EndKey: msstw.keyRanges[msstw.currRange].End.Key}
> ```

Add a comment about what we're doing here.
___
*[pkg/storage/store_snapshot.go, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-HPS9e3zflTeD2ZJ6B:-M-HPS9e3zflTeD2ZJ6C:b-7mtt8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef6d8591152aabc7421b07a35ff3522e0dae53ec/pkg/storage/store_snapshot.go#L154)):*
> ```Go
> 		return nil
> 	}
> 	// Only create file if the range is non-empty to avoid ingesting an empty SST later on since range dels are skipped for empty ranges.
> ```

Wrap this at 80 characters.
___
*[pkg/storage/store_snapshot.go, line 202 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-HPjgz-mWfPLp9D01D:-M-HPjh-1qZHE7S4HbeD:baqqc3m) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef6d8591152aabc7421b07a35ff3522e0dae53ec/pkg/storage/store_snapshot.go#L202)):*
> ```Go
> 		}
> 	}
> 	if err := msstw.maybeCreateNewFile(ctx); err != nil {
> ```

nit: I'd move this below the assertion. Also, wrap the error with `errors.Wrap`.
___
*[pkg/storage/rditer/replica_data_iter.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-HQBh7CgFfKH9F5Mct:-M-HQBh7CgFfKH9F5Mcu:b-9krvec) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef6d8591152aabc7421b07a35ff3522e0dae53ec/pkg/storage/rditer/replica_data_iter.go#L59)):*
> ```Go
> 	defer it.Close()
> 	it.SeekGE(engine.MakeMVCCMetadataKey(span.Key))
> 	if valid, _ := it.Valid(); !valid {
> ```

We should propagate these errors. If we hit the iterator bounds then we'll get `false, nil` from `Valid`, but if we hit a real error like RocksDB corruption, we'll want to infomr the caller.
___
*[pkg/storage/rditer/replica_data_iter_test.go, line 280 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-HQY6H1HR0uISc-fkK:-M-HQY6H1HR0uISc-fkL:b-8rvn0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef6d8591152aabc7421b07a35ff3522e0dae53ec/pkg/storage/rditer/replica_data_iter_test.go#L280)):*
> ```Go
> 
> 	span := roachpb.Span{
> 		Key: roachpb.Key("a"),
> ```

I think you need to `gofmt` this file.
___
*[pkg/storage/rditer/replica_data_iter_test.go, line 283 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-HSI7eBIGqw0f_63AA:-M-HSI7eBIGqw0f_63AB:bb003kz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef6d8591152aabc7421b07a35ff3522e0dae53ec/pkg/storage/rditer/replica_data_iter_test.go#L283)):*
> ```Go
> 		EndKey:   roachpb.Key("z"),
> 	}
> 	if _, empty := ConstrainToKeys(eng, span); !empty {
> ```

We've been slowly switching tests over to use the `require` package. We should use that here. Doing so will make it easy to assert that both return values are what we expect.
___
*[pkg/storage/rditer/replica_data_iter_test.go, line 313 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-HQaT69oXPTJQSJSh_:-M-HQaT69oXPTJQSJSha:b-yvhiug) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef6d8591152aabc7421b07a35ff3522e0dae53ec/pkg/storage/rditer/replica_data_iter_test.go#L313)):*
> ```Go
> 	key := keys.TransactionKey(roachpb.Key(desc.StartKey), uuid.MakeV4())
> 	ts := hlc.Timestamp{}
> 	if err := engine.MVCCPut(context.Background(), eng, nil, key, ts, roachpb.MakeValueFromString("value"), nil); err != nil {
> ```

This is still performing the Put here. Did you determine that it was needed?


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Another point here is that we generally try to write descriptive commit messages that inform readers of the goal of each change and the approach it took to accomplish that goal. Here's a resource that discusses how to write a good git commit message: https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages.

I know I mentioned that we should get this up to unblock people, so it's likely that that explains the temporary brevity. This is just something to keep in mind.

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44725)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @OwenQian)

___
*[pkg/storage/rditer/replica_data_iter.go, line 68 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-jSChc2_LfZplL8YTK:-M-jSChc2_LfZplL8YTL:b-dpyipr) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05f56a44aac312eff929e0caea73bbd4627e177/pkg/storage/rditer/replica_data_iter.go#L68)):*
> ```Go
> 	it.SeekLT(engine.MakeMVCCMetadataKey(span.EndKey))
> 	if valid, err := it.Valid(); !valid {
> 		return roachpb.Span{}, true, err
> ```

seems like in this case `err != nil` must be true, since we know there is a key that is in the span (or we have a bug somewhere). If yes, it may be worth confirming that here and explicitly creating an error if it is nil.


<!-- Sent from Reviewable.io -->

OwenQian commented:
___
*[pkg/storage/rditer/replica_data_iter.go, line 68 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-jSChc2_LfZplL8YTK:-M-k-Bme8KEcpe-boZ4w:bmnm00l) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05f56a44aac312eff929e0caea73bbd4627e177/pkg/storage/rditer/replica_data_iter.go#L68)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

seems like in this case `err != nil` must be true, since we know there is a key that is in the span (or we have a bug somewhere). If yes, it may be worth confirming that here and explicitly creating an error if it is nil.
</blockquote></details>

Added explicit error check for when the second iterator scan unexpected returns an error once we already know the span is non-empty.


<!-- Sent from Reviewable.io -->

OwenQian commented:
___
*[pkg/storage/rditer/replica_data_iter.go, line 68 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-jSChc2_LfZplL8YTK:-M13LSBk4q0on9MOB0ZZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05f56a44aac312eff929e0caea73bbd4627e177/pkg/storage/rditer/replica_data_iter.go#L68)):*
<details><summary><i>Previously, OwenQian (Owen Qian) wrote…</i></summary><blockquote>

Added explicit error check for when the second iterator scan unexpected returns an error once we already know the span is non-empty.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Commit:
storage: constrained span of rangedel in ClearRange to keys in range

Constrains the width of the range deletion tombstone to the span of keys
actually present within the range. If the range has no kv-entries, then skip
the rangedel completely.

Before this change, when receiving a snapshot, the original file
would have a range deletion tombstone that spanned the entire range written to
it regardless of the actual keys contained in the range or if the range was
empty. This resulted in the creation of excessively wide tombstones, which has
significant performance implications since the wide tombstones impede
compaction.

Fixes #44048.

Release note: None.

Update documentation.

Constrain width of range deletion tombstone for subsumed ranges.

Address nathan's reviews.

Address comments and fix pkg renaming issues with engine to storage.

OwenQian commented:
___
*[pkg/storage/rditer/replica_data_iter.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-HQBh7CgFfKH9F5Mct:-M17x1L94vBjx5X5c6hw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef6d8591152aabc7421b07a35ff3522e0dae53ec/pkg/storage/rditer/replica_data_iter.go#L59)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We should propagate these errors. If we hit the iterator bounds then we'll get `false, nil` from `Valid`, but if we hit a real error like RocksDB corruption, we'll want to infomr the caller.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

OwenQian commented:
___
*[pkg/storage/replica_sst_snapshot_storage_test.go, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-HPBps991KlDhDKVjJ:-M17ywYRCCEJHUAD1ybS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef6d8591152aabc7421b07a35ff3522e0dae53ec/pkg/storage/replica_sst_snapshot_storage_test.go#L52)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We'll need to update this comment. Then we'll want to test that once we start writing to the files, they get added to the slice.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

OwenQian commented:
___
*[pkg/storage/store_snapshot.go, line 110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-HPMPX05uySW-6M0aF:-M17z4lV0FW3PA07kFHk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef6d8591152aabc7421b07a35ff3522e0dae53ec/pkg/storage/store_snapshot.go#L110)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Add a period to the end of this.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

OwenQian commented:
___
*[pkg/storage/store_snapshot.go, line 149 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-HPVg_A81bB8KLn33v:-M17zYAm28SmlUsodpci:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef6d8591152aabc7421b07a35ff3522e0dae53ec/pkg/storage/store_snapshot.go#L149)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Add a comment about what we're doing here.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

OwenQian commented:
___
*[pkg/storage/rditer/replica_data_iter_test.go, line 280 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-HQY6H1HR0uISc-fkK:-M18-05MBcySQwR39RgK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef6d8591152aabc7421b07a35ff3522e0dae53ec/pkg/storage/rditer/replica_data_iter_test.go#L280)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I think you need to `gofmt` this file.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

OwenQian commented:
___
*[pkg/storage/rditer/replica_data_iter_test.go, line 313 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-HQaT69oXPTJQSJSh_:-M18-13-A2MaCY0ZTL2b:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef6d8591152aabc7421b07a35ff3522e0dae53ec/pkg/storage/rditer/replica_data_iter_test.go#L313)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is still performing the Put here. Did you determine that it was needed?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

OwenQian commented:
___
*[pkg/storage/rditer/replica_data_iter_test.go, line 283 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-HSI7eBIGqw0f_63AA:-M18-2lU3FtQigoPQdgE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef6d8591152aabc7421b07a35ff3522e0dae53ec/pkg/storage/rditer/replica_data_iter_test.go#L283)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We've been slowly switching tests over to use the `require` package. We should use that here. Doing so will make it easy to assert that both return values are what we expect.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

OwenQian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44725)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)

___
*[pkg/storage/store_snapshot.go, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-HPS9e3zflTeD2ZJ6B:-M17zeBV6GYjy-dvvRAK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef6d8591152aabc7421b07a35ff3522e0dae53ec/pkg/storage/store_snapshot.go#L154)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Wrap this at 80 characters.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

OwenQian commented:
___
*[pkg/storage/store_snapshot.go, line 202 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-HPjgz-mWfPLp9D01D:-M180Hiw1lCPuyN7Na0x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef6d8591152aabc7421b07a35ff3522e0dae53ec/pkg/storage/store_snapshot.go#L202)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: I'd move this below the assertion. Also, wrap the error with `errors.Wrap`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

nvb reviewed (changes requested):
This is looking a lot better. We still have some work to do to get this test solid, but we're getting closer.

> Reviewed 3 of 9 files at r7, 1 of 1 files at r8, 5 of 5 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-:-M1BoyV6B4ZymqtPCfEO:buowkd6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @OwenQian and @sumeerbhola)

___
*[pkg/storage/client_merge_test.go, line 2918 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1BmWhC6okviWItlc8F:-M1BmWhC6okviWItlc8G:bgvhh6s) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/client_merge_test.go#L2918)):*
> ```Go
> 			sstFile := &engine.MemFile{}
> 			sst := engine.MakeIngestionSSTWriter(sstFile)
> 			span, empty, err := rditer.ConstrainToKeys(sendingEng, roachpb.Span{Key: r.Start.Key, EndKey: r.End.Key})
> ```

This is similar to what we talked about before. We don't want to re-use the logic we're trying to test in the test itself - that's not a test of anything. If `ConstrainToKeys` was broken then we would never catch it here, right? This is similar to the following code:
```go
func plus(a, b int) int {
    // Wrong!
    return a * b
}

func TestPlus(t *testing.T) {
    const a, b = 10, 20
    exp := plus(a, b)
    // Will pass, even though plus is completely broken.
    require.Equal(t, exp, plus(a, b))
}
```

Instead of this, the test should know exactly what this clear range should encompass, and it should add that to the sst.

In fact, I'm realizing that we have the same issue with the rest of this SST. Instead of using the sender's engine to create it, we should be able to predict the contents of the SST ourselves - we were the ones who wrote the data it should contain in:
```
// ...
for _, key := range []roachpb.Key{roachpb.Key("a"), roachpb.Key("b"), roachpb.Key("c")} {
    if _, pErr := client.SendWrapped(ctx, distSender, incrementArgs(key, 1)); pErr != nil {
        t.Fatal(pErr)
    }
    mtc.waitForValues(key, []int64{1, 1, 1})
}

// ...
for i := 0; i < 10; i++ {
    key := roachpb.Key("d" + strconv.Itoa(i))
    if _, pErr := client.SendWrapped(ctx, distSender, incrementArgs(key, 1)); pErr != nil {
        t.Fatal(pErr)
    }
    mtc.waitForValues(key, []int64{1, 1, 1})
}
```
___
*[pkg/storage/replica_sst_snapshot_storage_test.go, line 137 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1Bic2p8L3tfX9NOy_D:-M1Bic2p8L3tfX9NOy_E:bbfkcdm) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/replica_sst_snapshot_storage_test.go#L137)):*
> ```Go
> 	sstSnapshotStorage := NewSSTSnapshotStorage(eng, testLimiter)
> 	var scratch *SSTSnapshotStorageScratch
> 	// Expect a SST file to be created for a range only if that range contains
> ```

Let's add a pair of subtests using `testutils.RunTrueAndFalse` to test with and without these keys. We also want to test that without these keys, the SSTs are not created.
___
*[pkg/storage/replica_sst_snapshot_storage_test.go, line 141 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1BjQAn7nTXuvLMKjhe:-M1BjQAn7nTXuvLMKjhf:b-4a08j4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/replica_sst_snapshot_storage_test.go#L141)):*
> ```Go
> 	// three types of replicated key ranges and check to see if the number of
> 	// SST files created matches what we expect.
> 	insertedKeys := []roachpb.Key{rangeIDLocalKey, rangeLocalKey, userKey}
> ```

This all seems a little fragile. We're trying to line up these keys with corresponding key ranges that are returned by `rditer.MakeReplicatedKeyRanges`. What if the implementation of `rditer.MakeReplicatedKeyRanges` changes? What if it returns keys in a different order?

I think we can improve this. Instead of creating these keys, can we just use the `.Start` key of each key range? Something like:
```
insertedKeys := make([]engine.MVCCKey, len(keyRanges))
for i, kr := range keyRanges {
    insertedKeys[i] = kr.Start
}
```
___
*[pkg/storage/replica_sst_snapshot_storage_test.go, line 142 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1BjpDbBWSE3xPJ-A8h:-M1BjpDbBWSE3xPJ-A8i:bf7503n) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/replica_sst_snapshot_storage_test.go#L142)):*
> ```Go
> 	// SST files created matches what we expect.
> 	insertedKeys := []roachpb.Key{rangeIDLocalKey, rangeLocalKey, userKey}
> 	for numberOfPuts := 0; numberOfPuts < 4; numberOfPuts++ {
> ```

Don't hard-code 4 here. It's a function of the length of `keyRanges`, right? Then let's use that.
___
*[pkg/storage/store_snapshot.go, line 110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-HPMPX05uySW-6M0aF:-M1Bklfx14wffbVn-bJk:b-6igy2g) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef6d8591152aabc7421b07a35ff3522e0dae53ec/pkg/storage/store_snapshot.go#L110)):*
<details><summary><i>Previously, OwenQian (Owen Qian) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

There are multiple range dels performed, so this comment needs to bit of work.
___
*[pkg/storage/store_snapshot.go, line 126 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1BkrtRE0DuIcVFe1i8:-M1BkrtRE0DuIcVFe1i9:b-hgteq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/store_snapshot.go#L126)):*
> ```Go
> 	// before flushing to disk.
> 	sstChunkSize int64
> 	// Used to scan for the Upper and LowerBound of ranges to restrict
> ```

s/Upper and LowerBound of ranges/minimum and maximum key in the range/
___
*[pkg/storage/store_snapshot.go, line 127 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1Bl1p90Z2O1loYoziX:-M1Bl1pA5N55DUyeGCHp:b-ak1wr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/store_snapshot.go#L127)):*
> ```Go
> 	sstChunkSize int64
> 	// Used to scan for the Upper and LowerBound of ranges to restrict
> 	// tombstone width.
> ```

s/tombstone/range deletion tombstone/
___
*[pkg/storage/store_snapshot.go, line 151 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1BlA5h6TByTSJvbC0j:-M1BlA5h6TByTSJvbC0k:b-3tif6r) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/store_snapshot.go#L151)):*
> ```Go
> 
> func (msstw *multiSSTWriter) initSST(ctx context.Context) error {
> 	span := roachpb.Span{Key: msstw.keyRanges[msstw.currRange].Start.Key, EndKey: msstw.keyRanges[msstw.currRange].End.Key}
> ```

Fold this line to look like:
```
span := roachpb.Span{
    Key: msstw.keyRanges[msstw.currRange].Start.Key, 
    EndKey: msstw.keyRanges[msstw.currRange].End.Key,
}
```
___
*[pkg/storage/store_snapshot.go, line 184 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1BlPwoEYsUVoFRe_FH:-M1BlPwoEYsUVoFRe_FI:b-t5chqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/store_snapshot.go#L184)):*
> ```Go
> }
> 
> // maybeCreateNewFile creates a new file if currSST is closed. Otherwise it is
> ```

nit: move this above `finalizeSST`.
___
*[pkg/storage/store_snapshot.go, line 187 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1BlVuF01BT4SgP5Mxz:-M1BlVuF01BT4SgP5My-:bw8btem) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/store_snapshot.go#L187)):*
> ```Go
> // a no-op. maybeCreateNewFile is idempotent.
> func (msstw *multiSSTWriter) maybeCreateNewFile(ctx context.Context) error {
> 	if !msstw.currSST.Closed() {
> ```

Add a small comment like:
```
// The file was already created.
```
___
*[pkg/storage/rditer/replica_data_iter.go, line 62 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1BlprA9X5yKhMW8a_l:-M1BlprA9X5yKhMW8a_m:b-zbekhn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/rditer/replica_data_iter.go#L62)):*
> ```Go
> 	defer it.Close()
> 	it.SeekGE(engine.MakeMVCCMetadataKey(span.Key))
> 	if valid, err := it.Valid(); !valid {
> ```

Split this into two cases like you have below.
___
*[pkg/storage/rditer/replica_data_iter.go, line 69 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1Blm4A27LLR-v_osyo:-M1Blm4A27LLR-v_osyp:b-hosrr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/rditer/replica_data_iter.go#L69)):*
> ```Go
> 	it.SeekLT(engine.MakeMVCCMetadataKey(span.EndKey))
> 	if valid, err := it.Valid(); err != nil {
> 		return roachpb.Span{}, true, errors.Wrapf(err, "Unexpected error when constraining non-empty span")
> ```

s/Unexpected/unexpected/
___
*[pkg/storage/rditer/replica_data_iter.go, line 71 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1Bm0RP7PBdhOWOzRhO:-M1Bm0RP7PBdhOWOzRhP:b-etlvyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/rditer/replica_data_iter.go#L71)):*
> ```Go
> 		return roachpb.Span{}, true, errors.Wrapf(err, "Unexpected error when constraining non-empty span")
> 	} else if !valid {
> 		return roachpb.Span{}, true, err
> ```

`return roachpb.Span{}, true, nil`
___
*[pkg/storage/rditer/replica_data_iter_test.go, line 303 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1BmEb6BggNaUrVfPTN:-M1BmEb6BggNaUrVfPTO:b-ijcdk3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/rditer/replica_data_iter_test.go#L303)):*
> ```Go
> 	require.False(t, empty)
> 	require.Equal(t, span.Key, insertedKey)
> 	require.Equal(t, span.EndKey, insertedKey.Next())
> ```

There are a few more cases to add here:
1. after adding a second key
2. after adding a key past the end of the search span
3. after adding a key before the start of the search span


<!-- Sent from Reviewable.io -->

OwenQian commented:
___
*[pkg/storage/client_merge_test.go, line 2918 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1BmWhC6okviWItlc8F:-M1CLIVrBtLS04pM_lmF:b-76yw5m) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/client_merge_test.go#L2918)):*
> This is similar to what we talked about before. We don't want to re-use the logic we're trying to test in the test itself - that's not a test of anything.

I might be confusing concepts here, but I'm curious about your thoughts on testing approach here:

Wouldn't a malfunction in ConstrainForKeys be caught in one of the unit tests for ConstrainForKeys e.g. TestConstrainToKeysNonEmptyRange, and so for the purpose of this integration test, we can assume that if those tests are passing, then ConstrainForKeys has the behavior we expect it to? Or is the preferred approach that we still want to minimize the number of assumptions that we make about underlying implementation even given the unit test?

I'm thinking if we used a separate code path to generate the expected SSTs that would couple the test tightly with the underlying implementation. Then if the behavior of ConstrainToKeys changed but we semantically still wanted to "constrain to keys", this test would break (along with the ConstrainToKeys unit tests).

I know we talked offline about the three different levels of testing and how CRDB wasn't big on mocking/unit tests. Sorry if I'm mixing up concepts here, I definitely do see how the current set-up is a bit tautological, so some extra clarity here would be much appreciated.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44725)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @OwenQian and @sumeerbhola)

___
*[pkg/storage/client_merge_test.go, line 2918 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1BmWhC6okviWItlc8F:-M1CTERy0w2sEOyvZt8T:b-8k9fc5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/client_merge_test.go#L2918)):*
> Wouldn't a malfunction in ConstrainForKeys be caught in one of the unit tests for ConstrainForKeys

Ideally, it would, yes. We use unit tests to ensure that a given piece of logic produces expected results over its input domain. But in addition to unit tests, we also want to make assertions across multiple layers to strengthen our confidence that each component is interacting correctly and that we weren't missing any unit tests. Testing at only a single layer leads to blind spots.

> Or is the preferred approach that we still want to minimize the number of assumptions that we make about underlying implementation even given the unit test?

Yes, to some degree. We certainly need to make assumptions when writing tests, but we also want to make sure we're making strong assertions which is only possible by avoiding assuming too much.

> I'm thinking if we used a separate code path to generate the expected SSTs that would couple the test tightly with the underlying implementation. 

We seem to be thinking about what constitutes "implementation details" differently. This test is verifying that the data ingested during a Range snapshot is what we expect it to be. So from the perspective of this test, the entries that are ingested are the entire point - not an implementation detail of the range snapshot mechanism. The same applies to this range deletion tombstone. From the perspective of the test, the bounds of the range deletion tombstone that is ingested is the important part, not that there is a range deletion tombstone that meets some abstract "constrained to keys" notion. Put another way, this test should fail if we changed the semantics of ConstrainToKeys.

Arguments could be made either way here though and there's not a hard rule on what should and shouldn't be assumed. I think the best advice I can give is that tests should convince their reader that the functionality they are testing is correct. Each assumption we make in tests works against this goal.


<!-- Sent from Reviewable.io -->

OwenQian commented:
___
*[pkg/storage/client_merge_test.go, line 2918 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1BmWhC6okviWItlc8F:-M1CLBAKAwCXvwhUHE_O:b-7ko672) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/client_merge_test.go#L2918)):*
> In fact, I'm realizing that we have the same issue with the rest of this SST. Instead of using the sender's engine to create it, we should be able to predict the contents of the SST ourselves - we were the ones who wrote the data it should contain in:

Updated test to generate expected SST from scratch rather than relying on the sender's engine and removed dependency on ConstrainToKeys.


<!-- Sent from Reviewable.io -->

OwenQian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44725)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)

___
*[pkg/storage/replica_sst_snapshot_storage_test.go, line 137 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1Bic2p8L3tfX9NOy_D:-M1R_v7O3yXDgyPRSMxy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/replica_sst_snapshot_storage_test.go#L137)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Let's add a pair of subtests using `testutils.RunTrueAndFalse` to test with and without these keys. We also want to test that without these keys, the SSTs are not created.
</blockquote></details>

Done.
___
*[pkg/storage/replica_sst_snapshot_storage_test.go, line 141 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1BjQAn7nTXuvLMKjhe:-M1CSsBZ1lf6ZtjpvyeG:b-faz0ia) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/replica_sst_snapshot_storage_test.go#L141)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This all seems a little fragile. We're trying to line up these keys with corresponding key ranges that are returned by `rditer.MakeReplicatedKeyRanges`. What if the implementation of `rditer.MakeReplicatedKeyRanges` changes? What if it returns keys in a different order?

I think we can improve this. Instead of creating these keys, can we just use the `.Start` key of each key range? Something like:
```
insertedKeys := make([]engine.MVCCKey, len(keyRanges))
for i, kr := range keyRanges {
    insertedKeys[i] = kr.Start
}
```
</blockquote></details>

Ooh, yeah I that a lot better, done.
___
*[pkg/storage/replica_sst_snapshot_storage_test.go, line 142 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1BjpDbBWSE3xPJ-A8h:-M1CRuC45mA14gvPJsXv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/replica_sst_snapshot_storage_test.go#L142)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Don't hard-code 4 here. It's a function of the length of `keyRanges`, right? Then let's use that.
</blockquote></details>

Done.
___
*[pkg/storage/store_snapshot.go, line 110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M-HPMPX05uySW-6M0aF:-M1CUQkD9eoJUnwRAbQK:b-phkiyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef6d8591152aabc7421b07a35ff3522e0dae53ec/pkg/storage/store_snapshot.go#L110)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

There are multiple range dels performed, so this comment needs to bit of work.
</blockquote></details>

Updated comment to add more details.
___
*[pkg/storage/store_snapshot.go, line 126 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1BkrtRE0DuIcVFe1i8:-M1CVjinA_EgdNteqXe0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/store_snapshot.go#L126)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

s/Upper and LowerBound of ranges/minimum and maximum key in the range/
</blockquote></details>

Done.
___
*[pkg/storage/store_snapshot.go, line 127 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1Bl1p90Z2O1loYoziX:-M1CVkJ7A68YikbJ9rOX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/store_snapshot.go#L127)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

s/tombstone/range deletion tombstone/
</blockquote></details>

Done.
___
*[pkg/storage/store_snapshot.go, line 151 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1BlA5h6TByTSJvbC0j:-M1CVuw178OvrjxCHba5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/store_snapshot.go#L151)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Fold this line to look like:
```
span := roachpb.Span{
    Key: msstw.keyRanges[msstw.currRange].Start.Key, 
    EndKey: msstw.keyRanges[msstw.currRange].End.Key,
}
```
</blockquote></details>

Done.
___
*[pkg/storage/store_snapshot.go, line 184 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1BlPwoEYsUVoFRe_FH:-M1CW6nt4OWnf1Szsxb5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/store_snapshot.go#L184)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: move this above `finalizeSST`.
</blockquote></details>

Done.
___
*[pkg/storage/store_snapshot.go, line 187 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1BlVuF01BT4SgP5Mxz:-M1CXKV0-YWvBg5IWLiI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/store_snapshot.go#L187)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Add a small comment like:
```
// The file was already created.
```
</blockquote></details>

Done.
___
*[pkg/storage/rditer/replica_data_iter.go, line 62 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1BlprA9X5yKhMW8a_l:-M1RapuUEdpOIundMxpl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/rditer/replica_data_iter.go#L62)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Split this into two cases like you have below.
</blockquote></details>

Done.
___
*[pkg/storage/rditer/replica_data_iter.go, line 69 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1Blm4A27LLR-v_osyo:-M1Ra4j0-xAI7_eWkYjv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/rditer/replica_data_iter.go#L69)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

s/Unexpected/unexpected/
</blockquote></details>

Done.
___
*[pkg/storage/rditer/replica_data_iter.go, line 71 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1Bm0RP7PBdhOWOzRhO:-M1RacuE72GblXVcBNZG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/rditer/replica_data_iter.go#L71)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`return roachpb.Span{}, true, nil`
</blockquote></details>

Done.
___
*[pkg/storage/rditer/replica_data_iter_test.go, line 303 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1BmEb6BggNaUrVfPTN:-M1RgPE5DD_rLBD1hN_9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a82b474b79bd0433e562482ccab6d2f716ba8c53/pkg/storage/rditer/replica_data_iter_test.go#L303)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

There are a few more cases to add here:
1. after adding a second key
2. after adding a key past the end of the search span
3. after adding a key before the start of the search span
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
This is looking significantly better. I think we'll be good to merge after addressing a few more small nits and stabilizing CI. Nice job!

> Reviewed 17 of 18 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-:-M1bw0T05jXtQ_Aoz85T:b-91zmzm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @OwenQian and @sumeerbhola)

___
*[pkg/kv/kvserver/client_merge_test.go, line 2854 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1buTTRAe-4Y5rmks8j:-M1buTTRAe-4Y5rmks8k:b-9z99a8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a29fe987a98f65f9a4f3739fca0cc9deb8fa8767/pkg/kv/kvserver/client_merge_test.go#L2854)):*
> ```Go
> }
> 
> // TestStoreRangeMergeRaftSnapshot sets up a situation which triggeres a
> ```

triggers
___
*[pkg/kv/kvserver/client_merge_test.go, line 2854 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1bufh43_zLYfULcQ-G:-M1bufh43_zLYfULcQ-H:b-fc6ebk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a29fe987a98f65f9a4f3739fca0cc9deb8fa8767/pkg/kv/kvserver/client_merge_test.go#L2854)):*
> ```Go
> }
> 
> // TestStoreRangeMergeRaftSnapshot sets up a situation which triggeres a
> ```

Nice comment!
___
*[pkg/kv/kvserver/client_merge_test.go, line 2880 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1buvGP96jmQFKxR1vE:-M1buvGP96jmQFKxR1vF:bqekrnq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a29fe987a98f65f9a4f3739fca0cc9deb8fa8767/pkg/kv/kvserver/client_merge_test.go#L2880)):*
> ```Go
> 	// Used to set the MVCCKey.Timestamp of the manually inserted keys so we
> 	// can generate a byte-by-byte equal expected SST without relying on the
> 	// sender for anything other than the timestamps.
> ```

We're not really relying on the sender for this, are we? We generated these keys.
___
*[pkg/kv/kvserver/client_merge_test.go, line 2881 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1buoEL3s8b3Gd7C2W9:-M1buoEL3s8b3Gd7C2WA:b-p3evtq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a29fe987a98f65f9a4f3739fca0cc9deb8fa8767/pkg/kv/kvserver/client_merge_test.go#L2881)):*
> ```Go
> 	// can generate a byte-by-byte equal expected SST without relying on the
> 	// sender for anything other than the timestamps.
> 	tsMap := make(map[string]hlc.Timestamp)
> ```

nit: rename this to something like `keyTimestamps`.
___
*[pkg/kv/kvserver/client_merge_test.go, line 2938 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1bv3fg-mZswCRO0Uuk:-M1bv3fg-mZswCRO0Uul:bxncfey) ([raw file](https://github.com/cockroachdb/cockroach/blob/a29fe987a98f65f9a4f3739fca0cc9deb8fa8767/pkg/kv/kvserver/client_merge_test.go#L2938)):*
> ```Go
> 			sst := engine.MakeIngestionSSTWriter(sstFile)
> 			// Expect the user KeyRange to be [a, d) but we only have keys at
> 			// a, b and c so expect the range del tombstone to span [a, c).
> ```

Reference ConstrainToKeys here.
___
*[pkg/kv/kvserver/client_merge_test.go, line 3048 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1bv_Gl2Z9L63bHrd0C:-M1bv_Gl2Z9L63bHrd0D:b-7u06dk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a29fe987a98f65f9a4f3739fca0cc9deb8fa8767/pkg/kv/kvserver/client_merge_test.go#L3048)):*
> ```Go
> 			t.Fatal(pErr)
> 		}
> 		tsMap[string(key)] = br.BatchResponse_Header.Timestamp
> ```

nit: `BatchResponse_Header` is embedded, so you can leave it out here.
___
*[pkg/kv/kvserver/rditer/replica_data_iter_test.go, line 288 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1bwIdD-GFbLsTnbSUU:-M1bwIdE4knu9r2t99vW:b1vaxln) ([raw file](https://github.com/cockroachdb/cockroach/blob/a29fe987a98f65f9a4f3739fca0cc9deb8fa8767/pkg/kv/kvserver/rditer/replica_data_iter_test.go#L288)):*
> ```Go
> 	defer eng.Close()
> 
> 	startKey := roachpb.Key("b")
> ```

nit: I think the test would actually be a lot easier to read if you inlined all of these variables. Want to try that out and see how it looks? I'll defer to your judgment.


<!-- Sent from Reviewable.io -->

OwenQian commented:
___
*[pkg/kv/kvserver/client_merge_test.go, line 2881 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1buoEL3s8b3Gd7C2W9:-M1fgvLH-p4UE7woGtHa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a29fe987a98f65f9a4f3739fca0cc9deb8fa8767/pkg/kv/kvserver/client_merge_test.go#L2881)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: rename this to something like `keyTimestamps`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

OwenQian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44725)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)

___
*[pkg/kv/kvserver/client_merge_test.go, line 2854 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1buTTRAe-4Y5rmks8j:-M1g9Baa0Idz6UwWY6c0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a29fe987a98f65f9a4f3739fca0cc9deb8fa8767/pkg/kv/kvserver/client_merge_test.go#L2854)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

triggers
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/client_merge_test.go, line 2854 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1bufh43_zLYfULcQ-G:-M1g9EUb9IMLT7ZOqwxx:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a29fe987a98f65f9a4f3739fca0cc9deb8fa8767/pkg/kv/kvserver/client_merge_test.go#L2854)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Nice comment!
</blockquote></details>

Thanks!
___
*[pkg/kv/kvserver/client_merge_test.go, line 2880 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1buvGP96jmQFKxR1vE:-M1g9WGvAZBdpowBrz3j:bq5wuc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a29fe987a98f65f9a4f3739fca0cc9deb8fa8767/pkg/kv/kvserver/client_merge_test.go#L2880)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We're not really relying on the sender for this, are we? We generated these keys.
</blockquote></details>

Updated comment.
___
*[pkg/kv/kvserver/client_merge_test.go, line 2938 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1bv3fg-mZswCRO0Uuk:-M1fhNwW6XsKdFw1YaUm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a29fe987a98f65f9a4f3739fca0cc9deb8fa8767/pkg/kv/kvserver/client_merge_test.go#L2938)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Reference ConstrainToKeys here.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/client_merge_test.go, line 3048 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1bv_Gl2Z9L63bHrd0C:-M1g4qhrALsSamSNouIx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a29fe987a98f65f9a4f3739fca0cc9deb8fa8767/pkg/kv/kvserver/client_merge_test.go#L3048)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: `BatchResponse_Header` is embedded, so you can leave it out here.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/rditer/replica_data_iter_test.go, line 288 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1bwIdD-GFbLsTnbSUU:-M1g7HOnCHTPyS7s-vgb:b8zzqjl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a29fe987a98f65f9a4f3739fca0cc9deb8fa8767/pkg/kv/kvserver/rditer/replica_data_iter_test.go#L288)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: I think the test would actually be a lot easier to read if you inlined all of these variables. Want to try that out and see how it looks? I'll defer to your judgment.
</blockquote></details>

I inlined the keys and added brief comments to explain why those keys were chosen.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> let's get this merged!

> Reviewed 6 of 6 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-:-M1gVCUL1lUn7XUMpw-r:bukag4l)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @OwenQian and @sumeerbhola)

___
*[pkg/kv/kvserver/client_merge_test.go, line 2939 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1gUwa77aa9QjbCHpGI:-M1gUwa77aa9QjbCHpGJ:b9qk1xl) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e59f5814d75e8bde64e083b402be31698e7d087/pkg/kv/kvserver/client_merge_test.go#L2939)):*
> ```Go
> 			// a, b and c so we expect rditer.ConstrainToKeys to restrict the
> 			// width of the range del tombstone to [a, c).
> 			if err := sst.ClearRange(storage.MakeMVCCMetadataKey(roachpb.Key("a")), storage.MakeMVCCMetadataKey(roachpb.Key("c").Next())); err != nil {
> ```

nit: wrap this line like:
```
if err := sst.ClearRange(
    storage.MakeMVCCMetadataKey(roachpb.Key("a")), 
    storage.MakeMVCCMetadataKey(roachpb.Key("c").Next()),
); err != nil {
```


<!-- Sent from Reviewable.io -->

OwenQian commented:
___
*[pkg/kv/kvserver/client_merge_test.go, line 2939 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/44725#-M1gUwa77aa9QjbCHpGI:-M1gVwOK4AMYh8A92wxs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e59f5814d75e8bde64e083b402be31698e7d087/pkg/kv/kvserver/client_merge_test.go#L2939)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: wrap this line like:
```
if err := sst.ClearRange(
    storage.MakeMVCCMetadataKey(roachpb.Key("a")), 
    storage.MakeMVCCMetadataKey(roachpb.Key("c").Next()),
); err != nil {
```
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

OwenQian commented:
bors r=nvanbenschoten
